distributed/actor.py

```diff
@@ -227,34 +253,61 @@ class ActorFuture:
     Actor
     """
 
-    def __init__(self, q, io_loop, result=None):
-        self.q = q
-        self.io_loop = io_loop
-        if result:
-            self._cached_result = result
-        self.status = "pending"
+    @abc.abstractmethod
+    def __await__(self):
+        pass
+
+    @abc.abstractmethod
+    def result(self, timeout=None):
+        pass
+
+    @abc.abstractmethod
+    def done(self):
+        pass
+
+    def __repr__(self):
+        return "<ActorFuture>"
+
+
+class _EagerActorFuture(ActorFuture):
+    """Future to an actor's method call when an actor calls another actor on the same worker"""
+
+    def __init__(self, result):
+        self._result = result
+
+    def __await__(self):
+        return self._result
+        yield
+
+    def result(self, timeout=None):
+        return self._result
+
+    def done(self):
+        return True
+
+
+class _ActorFuture(ActorFuture):
+    def __init__(self, io_loop):
+        self._io_loop = io_loop
+        self._event = _LateLoopEvent()
```

Comment:
I don't really have an objection to using an `Event` here, and it's not overly complex, but I did find myself wondering why this couldn't more simply wrap an `asyncio.Future()`. It seems like you could still get the late loop binding behavior by creating your future in `_result()`.

Reply:
if two or more tasks are both waiting on the same future and one task is cancelled then all the other tasks will be cancelled also

Reply:
I understand that the late-binding compatibility code is going away -- I'm still not sure I understand what you mean about the distinction in cancellation behavior. The snippet you wrote is what I had in mind, and it seems like it would work to me.

Reply:
here's a demo program of what happens when two tasks are waiting on the same future - and only one task is cancelled:

```python
import asyncio

class _ActorFuture:
    def __init__(self):
        self._fut = None

    def __await__(self):
        return self._result().__await__()

    def done(self):
        return self._fut.done()

    def _set_result(self, result):
        if self._fut is None:
            self._fut = asyncio.get_running_loop().create_future()
        try:
            self._fut.set_result(result)
        except asyncio.InvalidStateError:
            print("could not set result")

    async def _result(self):
        if self._fut is None:
            self._fut = asyncio.get_running_loop().create_future()

        out = await self._fut
        if out["status"] == "OK":
            return out["result"]
        raise out["exception"]


async def amain():
    af = _ActorFuture()

    async def async_fn():
        try:
            return await af
        except asyncio.CancelledError as e:
            print(f"{asyncio.current_task().get_name()} was cancelled")

    t1 = asyncio.create_task(async_fn(), name="task 1")
    t2 = asyncio.create_task(async_fn(), name="task 2")
    await asyncio.sleep(0)
    t1.cancel()
    af._set_result({"status": "OK", "result": 2})
    print(await asyncio.gather(t1, t2))

asyncio.run(amain())
```

it prints:

```
could not set result
task 1 was cancelled
task 2 was cancelled
[None, None]
```

task 2 gets cancelled as a consequence of task 1 being cancelled